src/levanter/models/llama.py

```diff
@@ -0,0 +1,279 @@
+        gate_proj = hnn.Linear.init(Out=Mlp, In=Embed, key=k_fc, use_bias=use_bias)
+        up_proj = hnn.Linear.init(Out=Mlp, In=Embed, key=k_up_proj, use_bias=use_bias)
+        down_proj = hnn.Linear.init(Out=Embed, In=Mlp, key=k_down_proj, use_bias=use_bias)
+        if isinstance(activation_fn, str):
+            activation_fn = ACT2FN[activation_fn]
+        act = activation_fn  # type: ignore
+        return LlamaMlp(gate_proj, up_proj, down_proj, act)
+
+    @named_call
+    def __call__(self, x: NamedArray) -> NamedArray:
+        hidden_states = self.gate_proj(x)
+        hidden_states = self.act(hidden_states)
+        hidden_states = hidden_states * self.up_proj(x)
+        outputs = self.down_proj(hidden_states)
+        return outputs
+
+
+class LlamaBlock(StateDictSerializationMixin, eqx.Module):
+    pass
+
+
+class LlamaTransformer(StateDictSerializationMixin, eqx.Module):
+    pass
+
+
+class LlamaEmbeddings(StateDictSerializationMixin, eqx.Module):
+    pass
+
+
+class LlamaLMHeadModel(eqx.Module):
+    pass
+
+
+class LlamaRotaryEmbedding(eqx.Module):
+    dim: int
+    max_position_embeddings: int = 2048
+    base: float = 10000
+    inv_freq: jnp.ndarray = eqx.static_field()
+    cos_cached: jnp.ndarray = eqx.static_field()
+    sin_cached: jnp.ndarray = eqx.static_field()
+    max_seq_len_cached: int = eqx.static_field()
+
+    def __init__(self, dim, max_position_embeddings=2048, base=10000):
+        self.dim = dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        self.inv_freq = 1.0 / (self.base ** (jnp.arange(0, self.dim, 2) / self.dim))
+        self.cos_cached, self.sin_cached = self._set_cos_sin_cache(seq_len=self.max_position_embeddings)
+
+    def _get_positional_ids(self):
+        """A helper function for the convenience of extending to two sub-classes
+        Here we use a standard positional encoding function, which was described in `Attention is all you need`.
+        """
+        return jnp.arange(self.max_seq_len_cached)
+
+    def _set_cos_sin_cache(self, seq_len):
+        self.max_seq_len_cached = seq_len
+        t = self._get_positional_ids()
+
+        # Evaluates the Einstein summation convention on the operands.
+        freqs = jnp.einsum("i,j->ij", t, self.inv_freq)
+        # Different from paper but following HF implementation
+        # It uses a different permutation in order to obtain the same calculation
+        emb = jnp.concatenate((freqs, freqs), axis=-1)
+        cos_cached = jnp.cos(emb)[None, None, :, :]
+        sin_cached = jnp.sin(emb)[None, None, :, :]
+        return cos_cached, sin_cached
+
+    def __call__(self, x, seq_len: int):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self.cos_cached, self.sin_cached = self._set_cos_sin_cache(seq_len=seq_len)
+
+        return (
+            self.cos_cached[:, :, :seq_len, ...],
```

Comment:
if you still do things this way (that is, supporting variable seq length with a cache), then:

`self.cos_cached[self.Pos, :seq_len]`